simplehub.go

```diff
@@ -53,37 +56,58 @@ type SimpleHub struct {
 //
 // The channel return value is closed when all the subscribers have been
 // notified of the event.
-func (h *SimpleHub) Publish(topic string, data interface{}) <-chan struct{} {
+func (h *SimpleHub) Publish(ctx context.Context, topic string, data interface{}) <-chan error {
 	h.mutex.Lock()
 	defer h.mutex.Unlock()
 
 	if prePublishTestHook != nil {
 		prePublishTestHook()
 	}
 
-	done := make(chan struct{})
-	wait := sync.WaitGroup{}
+	var wg sync.WaitGroup
+
+	done := make(chan error)
 
 	for _, s := range h.subscribers {
 		if s.topicMatcher(topic) {
-			wait.Add(1)
-			s.notify(
-				&handlerCallback{
-					topic: topic,
-					data:  data,
-					wg:    &wait,
-				})
+			wg.Add(1)
+			s.notify(&handlerCallback{
+				topic: topic,
+				data:  data,
+				wg:    &wg,
+			})
 		}
 	}
 
 	go func() {
-		wait.Wait()
+		// Allow the following waitgroup to unblock after a given set of time.
+		// If the notify never finishes, then this go routine will be around
+		// for a very long time and never close the done channel; leaking the
+		// gorountine.
+		if err := waitWithContext(ctx, &wg); err != nil {
```

Comment:
Not sure about this. If the caller doesn't accept the responsibility of of cancelling after a deadline, we're in the same boat, but now we've got 2 Goroutines for every one we had prior.